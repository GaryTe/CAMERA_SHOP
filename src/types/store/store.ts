import { store } from '../../store/store/store';
import { listCamera, CardProduct, Reviews, PromoProduct, Review } from '../types-response/types-response';

export type Cameras = {
  cameras: listCamera | null;
  modifiedListProduct: listCamera | null;
}

export type CardCamera = {
  dataCamera: CardProduct | null;
}

export type SimilaCameras = {
  dataSimilaCameras: listCamera | null;
  sortListCamerra: listCamera | null;
  counterMin: number;
  counterMax: number;
}

export type ReviewsCamera = {
  dataReviews: Reviews | Array<Review>;
  resultSortReviews: Reviews | null;
  reviews: Reviews | null;
  counter: number;
}

export type PromoCamera = {
  dataPromoCamera: PromoProduct | null;
}

export type Cart = {
  dataProduct: listCamera | Array<CardProduct>;
  loading: boolean;
}

export type RootState = ReturnType<typeof store.getState>;
