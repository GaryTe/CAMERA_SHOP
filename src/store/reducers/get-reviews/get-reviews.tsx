import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { fetchDataReviews, fetchPostNewReview } from '../../api-actions/api-actions';
import { sortReviews } from '../../../util/util';
import { ReviewsCamera } from '../../../types/store/store';
import { NameSpace } from '../../../const/const';
import { review, Review } from '../../../util/mocks';

const initialState: ReviewsCamera = {
  reviewsData: [],
  resultSortReviews: null,
  reviews: null,
  counter: 3
};

export const getReviews = createSlice({
  name: NameSpace.Reviews,
  initialState,
  reducers: {
    showMoreReviews: (state) => {
      if(state.resultSortReviews === null) {return;}
      state.counter += 3;
      state.reviews = state.resultSortReviews.slice(0, state.counter);
    },
    functionForTesting: (state, action: PayloadAction<Array<Review> | number>) => {
      state.reviewsData = typeof action?.payload === 'number' ? [review, review, review, review] : action.payload;
      state.reviews = [review, review, review];
    }
  },
  extraReducers(builder) {
    builder
      .addCase(fetchDataReviews.fulfilled, (state, action) => {
        state.reviewsData = action.payload;
        state.resultSortReviews = sortReviews(state.reviewsData);
        state.reviews = state.resultSortReviews.slice(0,state.counter);
      });
    builder
      .addCase(fetchPostNewReview.fulfilled, (state, action) => {
        state.reviewsData.push(action.payload);
        state.resultSortReviews = sortReviews(state.reviewsData);
        state.reviews = state.resultSortReviews.slice(0,state.counter);
      });
  }
});

export const {showMoreReviews, functionForTesting} = getReviews.actions;
