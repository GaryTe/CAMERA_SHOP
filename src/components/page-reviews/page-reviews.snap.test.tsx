import {render} from '@testing-library/react';
import {Provider} from 'react-redux';
import PegeReviews from './page-reviews';
import { store } from '../../store/store/store';
import { functionForTesting } from '../../store/reducers/get-reviews/get-reviews';

describe('Component: PegeReviews', () => {
  it('should render PegeReviews', () => {

    const {container} = render(
      <Provider store={store}>
        <PegeReviews cameraId={'1'} functionRequest={functionForTesting}/>
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
