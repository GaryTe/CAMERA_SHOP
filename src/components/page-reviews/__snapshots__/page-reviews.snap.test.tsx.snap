// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: PegeReviews should render PegeReviews 1`] = `
<div>
  <ul
    class="review-block__list"
  >
    <li
      class="review-card"
    >
      <div
        class="review-card__head"
      >
        <p
          class="title title--h4"
        >
          Кирилл
        </p>
        <time
          class="review-card__data"
          datetime="2022-04-13"
        >
          9 July
        </time>
      </div>
      <div
        class="rate review-card__rate"
      >
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <p
          class="visually-hidden"
        >
          Оценка: 
          5
        </p>
      </div>
      <ul
        class="review-card__list"
      >
        <li
          class="item-list"
        >
          <span
            class="item-list__title"
          >
            Достоинства:
          </span>
          <p
            class="item-list__text"
          >
            Легкая в плане веса, удобная в интерфейсе
          </p>
        </li>
        <li
          class="item-list"
        >
          <span
            class="item-list__title"
          >
            Недостатки:
          </span>
          <p
            class="item-list__text"
          >
            Быстро садиться зарядка
          </p>
        </li>
        <li
          class="item-list"
        >
          <span
            class="item-list__title"
          >
            Комментарий:
          </span>
          <p
            class="item-list__text"
          >
            Это моя первая камера. Я в восторге, нареканий нет
          </p>
        </li>
      </ul>
    </li>
    <li
      class="review-card"
    >
      <div
        class="review-card__head"
      >
        <p
          class="title title--h4"
        >
          Кирилл
        </p>
        <time
          class="review-card__data"
          datetime="2022-04-13"
        >
          9 July
        </time>
      </div>
      <div
        class="rate review-card__rate"
      >
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <p
          class="visually-hidden"
        >
          Оценка: 
          5
        </p>
      </div>
      <ul
        class="review-card__list"
      >
        <li
          class="item-list"
        >
          <span
            class="item-list__title"
          >
            Достоинства:
          </span>
          <p
            class="item-list__text"
          >
            Легкая в плане веса, удобная в интерфейсе
          </p>
        </li>
        <li
          class="item-list"
        >
          <span
            class="item-list__title"
          >
            Недостатки:
          </span>
          <p
            class="item-list__text"
          >
            Быстро садиться зарядка
          </p>
        </li>
        <li
          class="item-list"
        >
          <span
            class="item-list__title"
          >
            Комментарий:
          </span>
          <p
            class="item-list__text"
          >
            Это моя первая камера. Я в восторге, нареканий нет
          </p>
        </li>
      </ul>
    </li>
    <li
      class="review-card"
    >
      <div
        class="review-card__head"
      >
        <p
          class="title title--h4"
        >
          Кирилл
        </p>
        <time
          class="review-card__data"
          datetime="2022-04-13"
        >
          9 July
        </time>
      </div>
      <div
        class="rate review-card__rate"
      >
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <svg
          aria-hidden="true"
          height="16"
          width="17"
        >
          <use
            xlink:href="#icon-full-star"
          />
        </svg>
        <p
          class="visually-hidden"
        >
          Оценка: 
          5
        </p>
      </div>
      <ul
        class="review-card__list"
      >
        <li
          class="item-list"
        >
          <span
            class="item-list__title"
          >
            Достоинства:
          </span>
          <p
            class="item-list__text"
          >
            Легкая в плане веса, удобная в интерфейсе
          </p>
        </li>
        <li
          class="item-list"
        >
          <span
            class="item-list__title"
          >
            Недостатки:
          </span>
          <p
            class="item-list__text"
          >
            Быстро садиться зарядка
          </p>
        </li>
        <li
          class="item-list"
        >
          <span
            class="item-list__title"
          >
            Комментарий:
          </span>
          <p
            class="item-list__text"
          >
            Это моя первая камера. Я в восторге, нареканий нет
          </p>
        </li>
      </ul>
    </li>
  </ul>
  <div
    class="review-block__buttons"
  >
    <button
      class="btn btn--purple"
      type="button"
    >
      Показать больше отзывов
    </button>
  </div>
</div>
`;
